Source/WebCore/html/HTMLMediaElement.cpp

```diff
@@ -5676,7 +5676,7 @@ void HTMLMediaElement::mediaPlayerTimeChanged()
     double playbackRate = requestedPlaybackRate();
 
     // When the current playback position reaches the end of the media resource then the user agent must follow these steps:
-    if (dur && dur.isValid() && !dur.isPositiveInfinite() && !dur.isNegativeInfinite()) {
+    if ((dur || (!dur && !now)) && dur.isValid() && !dur.isPositiveInfinite() && !dur.isNegativeInfinite()) {
```

Comment:
Care you explain this?

Reply:
Yes. The test case provided by @emutavchi in https://github.com/WebPlatformForEmbedded/WPEWebKit/issues/1366 is expecting the "ended" event to finish the test. In this specific case, there are no buffers appended, so both duration and currentTime are zero. If you start playing in those circumstances, the video should just "end" immediately, but the previous version of the condition was preventing that by just ignoring the timeChanged notification (and bailing out by itself to be able to detect the "end" condition). This change avoids this.
I've checked [the spec](https://html.spec.whatwg.org/multipage/media.html#playing-the-media-resource) ("When the current playback position reaches the end of the media resource when the direction of playback is forwards, then the user agent must follow these steps: [...] 1. Fire an event named timeupdate at the media element.") and there's nothing preventing this new behaviour. I mean, there's nothing asking to avoid triggering the event if the duration is zero. Still, I tried to play safe and only get into the code block when the duration is zero and currentTime is also zero.

Reply:
What I want is why adding that check enables the notification.

Reply:
The endedEvent is only emitted from `HTMLMediaElement::mediaPlayerTimeChanged()`, [here](https://github.com/WebKit/WebKit/blob/9ebb270c66264c67a7b5bdd5f2629820b2a42321/Source/WebCore/html/HTMLMediaElement.cpp#L5704) and [here (but only for MediaStream)](https://github.com/WebKit/WebKit/blob/9ebb270c66264c67a7b5bdd5f2629820b2a42321/Source/WebCore/html/HTMLMediaElement.cpp#L5730). The first call is the one interesting in our case.
That call is inside the code block guarded by [the `if` condition that I'm modifying in this patch](https://github.com/WebKit/WebKit/blob/9ebb270c66264c67a7b5bdd5f2629820b2a42321/Source/WebCore/html/HTMLMediaElement.cpp#L5679). Without my modification, that code would never run and endedEvent would never be triggered. But it makes sense that a zero-length media "ends" if you try to play it, right?